adding a OV7670 driver.

jjsch-dev opened the pull request.

me-no-dev commented on Kconfig:
```diff
@@ -28,6 +35,17 @@ config SCCB_HARDWARE_I2C
         Enable this option if you want to use hardware I2C to control the camera.
         Disable this option to use software I2C.
 
+config ENABLE_TEST_PATTERN
```
I really do not see the point of having this setting here, given that you can just get the sensor and call the function in your code. This is a driver and not a final application. Please remove it.

me-no-dev commented on driver/camera.c:
```diff
@@ -1094,16 +1106,30 @@ esp_err_t camera_init(const camera_config_t* config)
+                s_state->dma_filter = &dma_filter_rgb888;
+            }
+            s_state->in_bytes_per_pixel = 2;       // camera sends RGB565
+            s_state->fb_bytes_per_pixel = 3;       // frame buffer stores RGB888
```
why? this uses more memory. You have functions to convert it after

me-no-dev commented on driver/private_include/camera_common.h:
```diff
@@ -3,7 +3,7 @@
 #include <stddef.h>
 #include <stdbool.h>
-#include "rom/lldesc.h"
+#include "esp32/rom/lldesc.h" 
```
why? it should work without `esp32/`. When I add support for the new chips, this header will not resolve.

me-no-dev replied:
ok, I did some research and here is how you should handle all headers that require `esp32/` change:
```cpp
#include "esp_system.h"
#ifdef ESP_IDF_VERSION_MAJOR // IDF 4+
#if CONFIG_IDF_TARGET_ESP32 // ESP32/PICO-D4
#include "esp32/rom/lldesc.h"
#endif
#else
#include "rom/lldesc.h" // ESP32 Before IDF 4.0
#endif
```

me-no-dev replied:
I actually took care of that, so revert the renames and rebase from master. Will make your PR cleaner too

me-no-dev commented on sensors/ov7670.c:
```diff
@@ -0,0 +1,438 @@
+	{HSYST, 0},
+    {HSYEN, 0},   
+
+    {MVFP, MVFP_MIRROR | MVFP_SUN}, 
```
if this is mirror and flip, I propose you remove them. You can set them in your code if your camera needs them

me-no-dev commented on sensors/ov7670.c:
```diff
@@ -0,0 +1,438 @@
+	{ RGB444,   0                           },  /* No RGB444 please */
+	{ COM1,     0                           },  /* CCIR601 */
+	{ COM15,    COM15_R00FF                 },
+    { MVFP,     /*MVFP_FLIP | */MVFP_MIRROR | MVFP_SUN}, /* Mirror and flip image */
```
see above. same comment

me-no-dev commented on sensors/ov7670.c:
```diff
@@ -0,0 +1,438 @@
+	{ COM1,     0                           },  /* CCIR601 */
+	{ COM15,    COM15_R00FF                 },
+    { MVFP,     /*MVFP_FLIP | */MVFP_MIRROR | MVFP_SUN}, /* Mirror and flip image */
+	{ COM9,     0x6A                        },  /* 128x gain ceiling; 0x8 is reserved bit */
```
gain ceiling is not part of the format. should be settable either in init or later though the API

me-no-dev commented on sensors/ov7670.c:
```diff
@@ -0,0 +1,438 @@
+	{ COM1,     0x0                         },	/* CCIR601 */
+	{ COM15,    COM15_RGB565 |COM15_R00FF   },
+    { MVFP,     MVFP_MIRROR | MVFP_SUN      },  /* Mirror image */ 
+	{ COM9,     0x6A                        }, 	/* 128x gain ceiling; 0x8 is reserved bit */
```
see above

me-no-dev commented on sensors/ov7670.c:
```diff
@@ -0,0 +1,438 @@
+    sensor->status.hmirror = 0;
+    sensor->status.vflip = 0;
+    sensor->status.colorbar = 0;
+    return 0;
```
why not read them from the sensor?

me-no-dev commented on driver/camera.c:
```diff
@@ -1094,16 +1106,30 @@ esp_err_t camera_init(const camera_config_t* config)
         }
         s_state->fb_bytes_per_pixel = 1;       // frame buffer stores Y8
     } else if (pix_format == PIXFORMAT_YUV422 || pix_format == PIXFORMAT_RGB565) {
-        s_state->fb_size = s_state->width * s_state->height * 2;
```
looking at the code, please revert this whole change. the camera will work fine without it. You should do the conversion to 888 in your code. Since a lot of the work happens in DMS Interrupts, you writing 50% more bytes can cause CPU starvation. In other words, I've seen it fully utilising the CPU inside interrupts.

jjsch-dev replied:
Hi, thanks for all suggestions, in the next days I will implements.

jjsch-dev commented:
I moved CONFIG_ENABLE_TEST_PATTERN, rewrote to correct the headers that require esp32, but the change to the to_bmp.c file is missing. Also delete the code in the camera.c of the RGB565.

me-no-dev commented:
I think you messed up things a bit :) the 5640 driver is in another branch. If you want to prep for that, I advise you to target that branch for now and not master.

jjsch-dev commented:
I'm sorry, I'm just starting to work with the GitHub repositories. Can you help me with an explanation to solve this issue?

me-no-dev commented:
you added the OV5640 driver and you should not have done it :) please revert those commits. If you really want to include and try the OV5640 driver, you need to target the ov5640 branch for your pull request and not master branch.

jjsch-dev commented:
ok, I'll reverse the last commits of the master. Is the correct way to work with the OV5640 sensor to create a new branch in my repository?

me-no-dev commented:
I would say, leave 5640 alone. I am working on it :)

me-no-dev commented:
https://github.com/espressif/esp32-camera/tree/ov5640/sensors

jjsch-dev commented:
@me-no-dev Today I have reverted the commits that conflicted with the master.

me-no-dev commented:
I fixed some issues. Please update your local copy and give it a shot :)

jjsch-dev commented:
Hi @me-no-dev, I updated my local copy (git pull origin master) but I don't see the changes. Sorry for my poor knowledge of github. Can you teach me how to update?

jjsch-dev commented:
Hi @me-no-dev Sorry for the audacity, but is it possible that you can manage with Espressif the shipment of a development kit of the ESP32-S2 to test the camera? In Argentina we are in quarantine so I have a lot of free time. Thanks

ShonP40 commented:
Any updates on this?

fuzun commented:
@jjsch-dev @me-no-dev What is the situation with this pull request? Is it tested?

jjsch-dev commented:
@fuzun  It is not tested with the latest version of IDF. When would you need to try a 7670?

fuzun commented:
@jjsch-dev I actually bought an OV2640 from a random Aliexpress seller but he sent me OV7670 instead. So I have to use it instead of OV2640. If your patch works fine, maybe I can use it. I can not test it right now so it would be the best if the maintainer of this repository tests it and merges it with the upstream.

jjsch-dev commented:
I can't promise, but this weekend I'm going to try to test the 7670 driver with the new IDF. I keep you informed.

Rajssss commented:
@jjsch-dev any update?

jjsch-dev commented:
Hello, today I started working, and passed the first step which was to test the hardware (I had disassembled it). In the next few days I will update the repository with the latest version of the IDF and esp32_camera to start testing.

Rajssss commented:
@jjsch-dev yay! Thanks.

jjsch-dev commented:
@Rajssss @fuzun today I Update my repository with the sensor code OV7670 to esp32_camera from 10/13/2020 and compile it with idf v4.3-dev-1197-g8bc19ba89-dirty.

![capture_13_10_2020_OV-7670](https://user-images.githubusercontent.com/55675185/95891074-f1360880-0d5a-11eb-972f-f878fb82d8ee.jpg)

fuzun commented:
@jjsch-dev Good news! Can you fix the conflicts as well? (Rebase your work on top of master)

Rajssss commented:
@jjsch-dev thanks man!

Rajssss commented:
@jjsch-dev can you tell me your module and pin map too?

jjsch-dev commented:
@Rajssss I give you the pin map that I use:

#
# ESP32 Camera Demo Configuration
#
CONFIG_XCLK_FREQ = 20000000

#
# Pin Configuration
#
CONFIG_D0 = 16
CONFIG_D1 = 25
CONFIG_D2 = 17
CONFIG_D3 = 33
CONFIG_D4 = 18
CONFIG_D5 = 32
CONFIG_D6 = 19
CONFIG_D7 = 35
CONFIG_XCLK = 5
CONFIG_PCLK = 14
CONFIG_VSYNC = 27
CONFIG_HREF = 23
CONFIG_SDA = 21
CONFIG_SCL = 22
CONFIG_RESET = 26
# end of Pin Configuration

Rajssss commented:
@jjsch-dev Thanks, which devkit you are using?

jjsch-dev commented:
ESP32-WROOM-32, but I don't know the exact version, please see the picture.
![esp32-wroom32-ov7670](https://user-images.githubusercontent.com/55675185/95895228-b7680080-0d60-11eb-9f0e-7a8d8b8008ef.jpeg)

Rajssss commented:
@jjsch-dev I am getting this error while building with idf v4.1
esp-idf-v4.1/components/esp32-camera/driver/twi.c: No such file or directory

I can't find the idf v4.3-dev-1197-g8bc19ba89-dirty to anywhere in the releases, can you tell me from where can I get it?

jjsch-dev commented:
> @jjsch-dev I am getting this error while building with idf v4.1
> esp-idf-v4.1/components/esp32-camera/driver/twi.c: No such file or directory
> 
> I can't find the idf v4.3-dev-1197-g8bc19ba89-dirty to anywhere in the releases, can you tell me from where can I get it?

@Rajssss twi.c was deleted because esp32_camera will not support i2c software emulation. The version 4.3 can be obtained by cloning the master of the [idf](https://github.com/espressif/esp-idf)

Rajssss commented:
@jjsch-dev I am getting this..
E (697) camera: Detected camera not supported.

E (697) camera: Camera probe failed with error 0x20004

E (697) camera_demo: Camera init failed with error 0x20004

Any idea?

jjsch-dev commented:
The 0x2004 indicates ESP_ERR_CAMERA_NOT_SUPPORTED, please check in menuconfig if you have enabled the OV 7670 sensor.
![menuconfig_enable_ov7670](https://user-images.githubusercontent.com/55675185/96023873-6f60e080-0e29-11eb-9957-ef9e8ec2b7ad.png)

Rajssss commented:
> 
> 
> The 0x2004 indicates ESP_ERR_CAMERA_NOT_SUPPORTED, please check in menuconfig if you have enabled the OV 7670 sensor.
> ![menuconfig_enable_ov7670](https://user-images.githubusercontent.com/55675185/96023873-6f60e080-0e29-11eb-9957-ef9e8ec2b7ad.png)

NVM fixed it, camera was not getting enough power due to bad jumper wires. Thanks its working fine now. 

can you tell me how can I setup a simple http webpage as camera view?

jjsch-dev commented:
@Rajssss good that you have been able to solve. Regarding a http server [esp_who](https://github.com/espressif/esp-who) is more complete than esp_camera_demo. Also, it is better to use a sensor that delivers images in jpg format.

Rajssss commented:
> @Rajssss good that you have been able to solve. Regarding a http server [esp_who](https://github.com/espressif/esp-who) is more complete than esp_camera_demo. Also, it is better to use a sensor that delivers images in jpg format.

Ok thanks will look into it. I am using a esp32-wroom-32d based board which doesn't have external PSRAM, it runs out every time when the buffer about 150KB, this happened when I tried JPEG mode. What do you suggest? Getting a better module is the only option?

Rajssss commented:
@jjsch-dev I have tried to build camera_webserver example from esp_who, but I am getting "camera: Detected camera not supported."
I have added CONFIG_OV7670_SUPPORT=y in sdkconfig, but still. Can you help me with that please.

jjsch-dev commented:
@Rajssss you're right, if you can, it really is a better option to use a module with PSRAM for image sensors and web interface. For some projects I use the ESP-32-CAM or the ESP-EYE..
![esp-eye](https://user-images.githubusercontent.com/55675185/96120356-c2d43c80-0ec4-11eb-9344-f4fe4e6546e1.jpeg)

jjsch-dev commented:
> @jjsch-dev I have tried to build camera_webserver example from esp_who, but I am getting "camera: Detected camera not supported."
> I have added CONFIG_OV7670_SUPPORT=y in sdkconfig, but still. Can you help me with that please.

my copy of esp-whoo is out of date, I'm going to download it and try to compile it for the OV-7670.

Rajssss commented:
> 
> 
> > @jjsch-dev I have tried to build camera_webserver example from esp_who, but I am getting "camera: Detected camera not supported."
> > I have added CONFIG_OV7670_SUPPORT=y in sdkconfig, but still. Can you help me with that please.
> 
> my copy of esp-whoo is out of date, I'm going to download it and try to compile it for the OV-7670.

Yeah I am still trying to make it work, but nothing happens after I press "start stream", Let me know if you update your fork please.

jjsch-dev commented:
@Rajssss Ok, I will do.

jjsch-dev commented:
@Rajssss I was able to get the camera_web_server example from the esp-whoo project to work with the OV-7670 connected to an ESP-32-WROM without PSRAM, by making the following changes. NOTE: for now it shows the controls of the OV2640.

1) Disable PSRAM in menuconfig.
![disable_psram](https://user-images.githubusercontent.com/55675185/96191538-a6fb8580-0f1a-11eb-8507-4e507e255663.png)

2) Modify the pixel format from JPG to RGB565, and the image size from UXVGA to QQVGA in the app_camera.c file
3) And finally, do not downgrade the framesize to QVGA, leave it in QQVGA.

```
    config.pixel_format = PIXFORMAT_RGB565; //PIXFORMAT_JPEG;
    //init with high specs to pre-allocate larger buffers
    config.frame_size = FRAMESIZE_QQVGA; //FRAMESIZE_UXGA;
    config.jpeg_quality = 10;
    config.fb_count = 2;

    // camera init
    esp_err_t err = esp_camera_init(&config);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Camera init failed with error 0x%x", err);
        return;
    }

    sensor_t * s = esp_camera_sensor_get();
    //initial sensors are flipped vertically and colors are a bit saturated
    if (s->id.PID == OV3660_PID) {
        s->set_vflip(s, 1);//flip it back
        s->set_brightness(s, 1);//up the blightness just a bit
        s->set_saturation(s, -2);//lower the saturation
    }
    //drop down frame size for higher initial frame rate
    s->set_framesize(s, FRAMESIZE_QQVGA); // FRAMESIZE_QVGA);
```
If the changes don't work for you, I'm working on modifying esp-who in my github repository with the changes for the OV7670, but I don't know when it will be available.

![esp-who](https://user-images.githubusercontent.com/55675185/96191672-f772e300-0f1a-11eb-930e-f8edc08a4c4f.gif)

Rajssss commented:
@jjsch-dev yeah I did exactly the same, but idk why it's not working for me... I will wait no problem.

Rajssss commented:
@jjsch-dev I am trying again today, and looks like I can get still images by clicking on "Get Still" on the webpage, but when I click on "Start Stream" It doesn't work. Here is the logs. If you can please check these out. https://del.dog/esplog.txt

jjsch-dev commented:
@Rajssss I cloned  [esp-who](https://github.com/jjsch-dev/esp-who/tree/support-OV7670)  in my repository, and modified the camera_web_server example to support the OV-7670. When the PSRAN chip is not available, the internal RAM is scarce, so in addition to reducing the resolution of the image, it is necessary to use dynamic buffers for wifi tx CONFIG_ESP32_WIFI_DYNAMIC_TX_BUFFER=y. I send you my sdconfig to use as an example. [sdkconfig_7670_NOT_PSRAM.zip](https://github.com/espressif/esp32-camera/files/5396780/sdkconfig_7670_NOT_PSRAM.zip)

If everything works fine, you should have a website like the following.
![esp-who_7670](https://user-images.githubusercontent.com/55675185/96354234-b8739780-10aa-11eb-9210-579ebe32fcab.gif)

jjsch-dev commented:
> @jjsch-dev I am trying again today, and looks like I can get still images by clicking on "Get Still" on the webpage, but when I click on "Start Stream" It doesn't work. Here is the logs. If you can please check these out. https://del.dog/esplog.txt

@Rajssss  When the PSRAN chip is not available, the internal RAM is scarce, so in addition to reducing the resolution of the image, you have to use dynamic buffers for wifi tx

# CONFIG_ESP32_WIFI_STATIC_TX_BUFFER is not set
CONFIG_ESP32_WIFI_DYNAMIC_TX_BUFFER = y
CONFIG_ESP32_WIFI_TX_BUFFER_TYPE = 1
CONFIG_ESP32_WIFI_DYNAMIC_TX_BUFFER_NUM = 32

Rajssss commented:
@jjsch-dev thanks, will try Today. Have you tried face recognition with that resolution? Also is it somehow possible to use any higher resolution than QQVGA?

jjsch-dev commented:
@Rajssss for these boards without PSRAM it is not possible because image recognition requires a large amount of memory.

The resolution is also limited by the available RAM, you can double it if instead of RGB you use grayscale.

Rajssss commented:
@jjsch-dev hello again, thanks for all the help you have provided. Now that I have bought a ESP32 DevKitC with WROVER module, 8MB of PSRAM, I want to get the best out of this OV7670 Cam, I see the driver doesn't support JPEG? so what is the max resolution that I can get?

jjsch-dev commented:
@Rajssss  hello, it is a good choice to increase the RAM to 8Mbytes because you will be able to use the maximum resolution of the sensor which is VGA (640x480).

Rajssss commented:
> 
> 
> @Rajssss hello, it is a good choice to increase the RAM to 8Mbytes because you will be able to use the maximum resolution of the sensor which is VGA (640x480).

Well, time for a better camera maybe? thinking of a OV5640.

jjsch-dev commented:
If the budget supports it, it is definitely better to use a camera like the OV5640.

Rajssss commented:
> 
> 
> If the budget supports it, it is definitely better to use a camera like the OV5640.

Thinking about this one, its OV5640D, should be supported right?
https://www.arducam.com/product/5-mp-ov5640d-cmos-af-camera-module-1-4-inch-module/

jjsch-dev commented:
@Rajssss  The module contains a supported sensor, but before buying check the pin-out and the supply voltages.

Rajssss commented:
@jjsch-dev Hello again, I am using a ESP32 DevKit V4 with WROVER module. Everything works fine in QQVGA but when I tried with QVGA I am getting this error:
E (43599) camera: Failed to get the frame on time!
E (43599) camera_httpd: Camera capture failed
W (43599) httpd_txrx: httpd_resp_send_err: 500 Internal Server Error - Server has encountered an unexpected error
W (43609) httpd_uri: httpd_uri: uri handler execution failed

In QQVGA it uses the internal ram only, but for the QVGA it uses the PSARM, maybe something wrong with PSRAM?

jjsch-dev commented:
@Rajssss Hello, the log is from the esp-who program? and for vga you have to enable psram.

Rajssss commented:
> @Rajssss Hello, the log is from the esp-who program? and for vga you have to enable psram.

Yes it's esp who, camera webserver example, and I have enabled the PSRAM. Actually it was enabled by default.

jjsch-dev commented:
@Rajssss, sorry for the late reply, in the next few days I will try to test this board.

Rajssss commented:
> @Rajssss, sorry for the late reply, in the next few days I will try to test this board.

Yup, no problem take your time.

jjsch-dev commented:
@Rajssss hi, could you tell me what is the speed of the camera clock?

Rajssss commented:
> 
> 
> @Rajssss hi, could you tell me what is the speed of the camera clock?

XCLK is 5MHz.

Rajssss commented:
> 
> 
> > @Rajssss hi, could you tell me what is the speed of the camera clock?
> 
> XCLK is 5MHz. PCLK is 14MHz. its all default.

You can find my sdkconfig along with whole repo here: https://github.com/Rajssss/esp-who_wrover/blob/master/examples/single_chip/camera_web_server/sdkconfig#L162

jjsch-dev commented:
please change the speed to 10 and 14 Mhz in this line (config.xclk_freq_hz = 20000000;) of the app_camera.c module

Rajssss commented:
> please change the speed to 10 and 14 Mhz in this line (config.xclk_freq_hz = 20000000;) of the app_camera.c module

You mean I should change XCLK to 10MHz instead of 5MHz?

jjsch-dev commented:
no, change this line to: config.xclk_freq_hz = 14000000; or config.xclk_freq_hz = 10000000;

Rajssss commented:
> 
> 
> no, change this line to: config.xclk_freq_hz = 14000000; or config.xclk_freq_hz = 10000000;

oh, ok.

Rajssss commented:
> 
> 
> no, change this line to: config.xclk_freq_hz = 14000000; or config.xclk_freq_hz = 10000000;

I have chnged it to 14MHz, but the app is crashing now.

ELF file SHA256: a1b54fed9983e4d3

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:4
load:0x3fff0034,len:5840
ho 0 tail 12 room 4
load:0x40078000,len:14988
load:0x40080400,len:4028
entry 0x40080648
I (922) psram: This chip is ESP32-D0WD
I (923) spiram: Found 64MBit SPI RAM device
I (923) spiram: SPI RAM mode: flash 80m sram 40m
I (926) spiram: PSRAM initialized, cache is in low/high (2-core) mode.
I (933) cpu_start: Pro cpu up.
I (937) cpu_start: Starting app cpu, entry point is 0x4008156c
I (928) cpu_start: App cpu up.
I (1824) spiram: SPI SRAM memory test OK
I (1832) cpu_start: Pro cpu start user code
I (1832) cpu_start: cpu freq: 240000000
I (1832) cpu_start: Application information:
I (1835) cpu_start: Project name:     camera_web_server
I (1841) cpu_start: App version:      v0.9.3-62-g74a5b30-dirty
I (1847) cpu_start: Compile time:     Nov 22 2020 05:39:23
I (1853) cpu_start: ELF file SHA256:  a1b54fed9983e4d3...
I (1859) cpu_start: ESP-IDF:          178b122
I (1865) heap_init: Initializing. RAM available for dynamic allocation:
I (1872) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (1878) heap_init: At 3FFBC3E8 len 00023C18 (143 KiB): DRAM
I (1884) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (1891) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (1897) heap_init: At 4009B808 len 000047F8 (17 KiB): IRAM
I (1903) spiram: Adding pool of 4096K of external SPI memory to heap allocator
I (1912) spi_flash: detected chip: generic
I (1916) spi_flash: flash io: qio
I (1921) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (2015) wifi:wifi driver task: 3ffc4f68, prio:23, stack:6656, core=0
I (2015) system_api: Base MAC address is not set
I (2015) system_api: read default base MAC address from EFUSE
I (2035) wifi:wifi firmware version: b9e8721
I (2035) wifi:wifi certification version: v7.0
I (2035) wifi:config NVS flash: enabled
I (2035) wifi:config nano formating: disabled
I (2035) wifi:Init data frame dynamic rx buffer num: 32
I (2045) wifi:Init management frame dynamic rx buffer num: 32
I (2045) wifi:Init management short buffer num: 32
I (2055) wifi:Init static tx buffer num: 16
I (2055) wifi:Init tx cache buffer num: 32
I (2055) wifi:Init static rx buffer size: 1600
I (2065) wifi:Init static rx buffer num: 10
I (2065) wifi:Init dynamic rx buffer num: 32
I (2075) wifi_init: rx ba win: 6
I (2075) wifi_init: tcpip mbox: 32
I (2075) wifi_init: udp mbox: 6
I (2085) wifi_init: tcp mbox: 6
I (2085) wifi_init: tcp tx win: 5744
I (2095) wifi_init: tcp rx win: 5744
I (2095) wifi_init: tcp mss: 1440
I (2095) wifi_init: WiFi IRAM OP enabled
I (2105) wifi_init: WiFi RX IRAM OP enabled
I (2115) camera wifi: wifi_init_softap finished.SSID:ESP32-Camera password:
I (2195) phy: phy_version: 4500, 0cd6843, Sep 17 2020, 15:37:07, 0, 0
I (2195) wifi:mode : softAP (84:0d:8e:e3:b7:49)
I (2195) wifi:Total power save buffer number: 8
I (2205) wifi:Init max length of beacon: 752/752
I (2205) wifi:Init max length of beacon: 752/752
I (2205) wifi:Set ps type: 0

I (2215) gpio: GPIO[14]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2225) gpio: GPIO[16]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2235) gpio: GPIO[17]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2235) gpio: GPIO[18]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2245) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2255) gpio: GPIO[23]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2265) gpio: GPIO[25]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2275) gpio: GPIO[27]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2285) gpio: GPIO[32]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2295) gpio: GPIO[33]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2305) gpio: GPIO[35]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (2315) sccb: pin_sda 21 pin_scl 22


abort() was called at PC 0x40084dfe on core 0

Backtrace:0x4008ddae:0x3ffbe450 0x4008e5dd:0x3ffbe470 0x40095a46:0x3ffbe490 0x40084dfe:0x3ffbe500 0x40084f5d:0x3ffbe530 0x40161239:0x3ffbe550 0x4016462f:0x3ffbe860 0x4016d20b:0x3ffbe890 0x400943bb:0x3ffbe8c0 0x4013d251:0x3ffbe910 0x400e54da:0x3ffbe940 0x400e4b8b:0x3ffbe990 0x400e527d:0x3ffbe9d0 0x400db2b9:0x3ffbea00 0x400daf7e:0x3ffbea80 0x4016cd1b:0x3ffbeaa0 0x40090ee1:0x3ffbeac0

jjsch-dev commented:
@Rajssss, hello I was analyzing the code, and I did not find what could be happening. As I am now using the OV7670 sensor with an ESP32-WROOM-32D, (which does not have PSRAM), I am going to buy an ESP32-WROVER- to perform tests that match your setup. In a few days I will tell you the result.

Rajssss commented:
> @Rajssss, hello I was analyzing the code, and I did not find what could be happening. As I am now using the OV7670 sensor with an ESP32-WROOM-32D, (which does not have PSRAM), I am going to buy an ESP32-WROVER- to perform tests that match your setup. In a few days I will tell you the result.

Great, thanks man :)

Rajssss commented:
@jjsch-dev hey bro any update?

jjsch-dev commented:
@Rajssss, Hello, last week I bought the module, I hope to have it in my possession this week.

Rajssss commented:
> @Rajssss, Hello, last week I bought the module, I hope to have it in my possession this week.

Great, I can help you on testing though, we should move our chat to a different platform though...

jjsch-dev commented:
@Rajssss  hello, today I finished changing the esp32-wroom-32d module for the esp32-wrover-e with that has 16 Mbytes of PSRAM, tomorrow I will start testing I keep you informed.

![IMG_20201211_122750](https://user-images.githubusercontent.com/55675185/101965788-9a329f80-3bf4-11eb-9d1f-5ea0d469e145.jpg)
![IMG_20201211_122855](https://user-images.githubusercontent.com/55675185/101965789-9acb3600-3bf4-11eb-985c-8c26abda0223.jpg)
![IMG_20201211_110808](https://user-images.githubusercontent.com/55675185/101965790-9acb3600-3bf4-11eb-9f24-831695f34cbf.jpg)
![IMG_20201211_141202](https://user-images.githubusercontent.com/55675185/101965791-9b63cc80-3bf4-11eb-93df-243003e0556e.jpg)
![IMG_20201211_141212](https://user-images.githubusercontent.com/55675185/101965794-9c94f980-3bf4-11eb-9420-f06cd50a6cc9.jpg)

Rajssss commented:
@jjsch-dev wew, cool!

jjsch-dev commented:
@Rajssss  hello, after adjusting the camera pins for the wrover module I was able to test with PSRAM and resolutions higher than QQVGA, which confirmed the problem you reported. I am working to solve it, as soon as I have more information I will contact you.

Rajssss commented:
> @Rajssss hello, after adjusting the camera pins for the wrover module I was able to test with PSRAM and resolutions higher than QQVGA, which confirmed the problem you reported. I am working to solve it, as soon as I have more information I will contact you.

That's some great news, let me know if I can help you somehow, though I still don't know how to setup a debugger with esp or esp IDF. It will be very helpful.

jjsch-dev commented:
Thanks for the help offered, I will take it into account. I have not used the debugger with esp32 either, for now I use the IDF log to obtain the information of the registers and code status.

Rajssss commented:
> Thanks for the help offered, I will take it into account. I have not used the debugger with esp32 either, for now I use the IDF log to obtain the information of the registers and code status.

I will get a cheap clone of esp-prog, this will help for sure.

jjsch-dev commented:
@Rajssss  Hello, sorry for the delay but the solution was more complex than I expected, by mistake I had the routines that are in charge of attending the ISR reading the image incorrectly set. I also had to modify the windowing (start/stop) of all formats and encountered some problems in the CIF format and in a conversion routine. Tomorrow I hope to upload the code for you to test it.

jjsch-dev commented:
@Rajssss hello, today I uploaded a version of [esp32-camera](https://github.com/jjsch-dev/esp32-camera) to my repository with the modifications that solve the QVGA problem and also, as I mentioned, it repairs inconsistencies with the rest of the formats.

Here are some GIF captures of esp-who in webcam mode with the OV-7670.

**VGA**
![ov7670_vga](https://user-images.githubusercontent.com/55675185/102813827-4704dd80-43a8-11eb-8a5f-bd5c7c680080.gif)

**QVGA**
![ov7670_qvga](https://user-images.githubusercontent.com/55675185/102813897-6bf95080-43a8-11eb-9f66-31e15376d7b0.gif)

**QQVGA**
![ov7670_qqvga](https://user-images.githubusercontent.com/55675185/102813971-8df2d300-43a8-11eb-8a0a-e4a68fcbae40.gif)

Rajssss commented:
> @Rajssss hello, today I uploaded a version of [esp32-camera](https://github.com/jjsch-dev/esp32-camera) to my repository with the modifications that solve the QVGA problem and also, as I mentioned, it repairs inconsistencies with the rest of the formats.
> 
> Here are some GIF captures of esp-who in webcam mode with the OV-7670.
> 
> **VGA**
> ![ov7670_vga](https://user-images.githubusercontent.com/55675185/102813827-4704dd80-43a8-11eb-8a5f-bd5c7c680080.gif)
> 
> **QVGA**
> ![ov7670_qvga](https://user-images.githubusercontent.com/55675185/102813897-6bf95080-43a8-11eb-9f66-31e15376d7b0.gif)
> 
> **QQVGA**
> ![ov7670_qqvga](https://user-images.githubusercontent.com/55675185/102813971-8df2d300-43a8-11eb-8a0a-e4a68fcbae40.gif)

Thanks bro, have you tested face recognition too?

jjsch-dev commented:
> Thanks bro, have you tested face recognition too?

it works, it is slower than with cameras with native jpg support, but it works fine.

Could you check if the update I make works with your device?

Rajssss commented:
> > Thanks bro, have you tested face recognition too?
> 
> it works, it is slower than with cameras with native jpg support, but it works fine.
> 
> Could you check if the update I make works with your device?

Ok, I will test and update you by today itself.

Rajssss commented:
@jjsch-dev thanks brother, just tested everything on QVGA resolution, everything works fine.

jjsch-dev commented:
@Rajssss thanks for the feedback, the resolution change can only be done in jpg format, because for YUV, RGB, it is necessary to restart the camera driver to modify the sizes of the internal buffers.
